cmd/logmonitor/main.go

```diff
@@ -0,0 +1,186 @@
+	databaseURI          = flag.String("database_uri", "", "The database connection URI")
+	waitBetweenDBUpdates = flag.Duration("database_updates", 5*time.Second, "How long to wait between database updates")
+
+	configFile = flag.String("config", "", "Config file containing flags, file contents can be overridden by command line flags")
+)
+
+// Errors
+var (
+	errServerMissing       = errors.New("server is missing: use the -server flag to set it")
+	errLogIDMissing        = errors.New("log ID is missing: use the -log_id flag to set it")
+	errPublicKeyMissing    = errors.New("the Trillian log public key path is missing: use the -log_public_key flag to set it")
+	errDatabaseMissing     = fmt.Errorf("database is missing: use the -database flag to set it to one of: %v", strings.Join(supportedDatabases, ", "))
+	errUnsupportedDatabase = fmt.Errorf("selected database set by the -database flag is unsupported: use one of: %v", strings.Join(supportedDatabases, ", "))
+)
+
+func verifyFlags() error {
+	if *serverAddr == "" {
+		return errServerMissing
+	}
+
+	if *logID == 0 {
+		return errLogIDMissing
+	}
+
+	if *logPublicKeyFile == "" {
+		return errPublicKeyMissing
+	}
+
+	if *database == "" {
+		return errDatabaseMissing
+	}
+
+	databaseIsSupported := false
+	for _, d := range supportedDatabases {
+		if *database == d {
+			databaseIsSupported = true
+			break
+		}
+	}
+	if !databaseIsSupported {
+		return errUnsupportedDatabase
+	}
+
+	return nil
+}
+
+func newLogMonitorFromFlags() (*LogMonitor, error) {
+	if err := verifyFlags(); err != nil {
+		return nil, err
+	}
+
+	db, err := Open(*database, *databaseURI)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := Setup(db); err != nil {
+		_ = db.Close()
+		return nil, err
+	}
+
+	startingRoot, err := GetCurrentRootFor(db, *logID)
+	if err != nil {
+		_ = db.Close()
+		return nil, err
+	}
+
+	logClient, err := newLogClientFromFlags(*startingRoot)
+	if err != nil {
+		_ = db.Close()
+		return nil, err
+	}
+
+	monitor := &LogMonitor{
+		logID:                *logID,
+		logClient:            logClient,
+		db:                   db,
+		waitBetweenDBUpdates: *waitBetweenDBUpdates,
+	}
+	return monitor, nil
+}
+
+func newLogClientFromFlags(startingRoot types.LogRootV1) (*client.LogClient, error) {
+	publicKey, err := pem.ReadPublicKeyFile(*logPublicKeyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	dialOpts, err := rpcflags.NewClientDialOptionsFromFlags()
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine dial options: %v", err)
+	}
+
+	conn, err := grpc.Dial(*serverAddr, dialOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial %s: %v", *serverAddr, err)
+	}
+
+	grpcClient := trillian.NewTrillianLogClient(conn)
+	hasher := objhasher.NewLogHasher(rfc6962.New(crypto.SHA256))
+	verifier := client.NewLogVerifier(hasher, publicKey, crypto.SHA256)
+	return client.New(*logID, grpcClient, verifier, startingRoot), nil
+}
+
+func runMonitor(ctx context.Context) error {
+	flag.Parse()
```

Comment:
`flag.Parse()` typically lives in `main()`

Reply:
I'm leaving it here to simplify testing. I basically call `runMonitor` in my tests instead of running main directly.

Reply:
... and the reason I don't want to call `main` directly in my tests is that it calls `glog.Exit` which would kill my test process.